Doppler.Sap.Job/DopplerCurrencyService/DopplerCurrencyService.cs

```diff
@@ -1,6 +1,82 @@
-﻿namespace Doppler.Sap.Job.Service.DopplerCurrencyService
+﻿using System;
+using System.Collections.Generic;
+using System.Net.Http;
+using System.Threading.Tasks;
+using Doppler.Sap.Job.Service.Dtos;
+using Doppler.Sap.Job.Service.Logger;
+using Doppler.Sap.Job.Service.Settings;
+using Newtonsoft.Json;
+
+namespace Doppler.Sap.Job.Service.DopplerCurrencyService
 {
-    public class DopplerCurrencyService
+    public class DopplerCurrencyService : IDopplerCurrencyService
     {
+        private readonly HttpClient _httpClient;
+        private readonly DopplerCurrencySettings _dopplerCurrencySettings;
+        private readonly ILoggerAdapter<DopplerCurrencyService> _logger;
+
+        public DopplerCurrencyService(
+            IHttpClientFactory httpClientFactory,
+            HttpClientPoliciesSettings httpClientPoliciesSettings,
+            DopplerCurrencySettings dopplerCurrencySettings,
+            ILoggerAdapter<DopplerCurrencyService> logger)
+        {
+            _httpClient = httpClientFactory.CreateClient(httpClientPoliciesSettings.ClientName);
+            _dopplerCurrencySettings = dopplerCurrencySettings;
+            _logger = logger;
+        }
+
+        public async Task<IList<CurrencyDto>> GetCurrencyByCode()
+        {
+            var dateTimeNow = DateTime.UtcNow;
+            var returnList = new List<CurrencyDto>();
+            
+            foreach (var currencyCode in _dopplerCurrencySettings.CurrencyCodeList)
+            {
+                var uri = new Uri(_dopplerCurrencySettings.Url + $"{currencyCode}/{dateTimeNow.Year}-{dateTimeNow.Month}-{dateTimeNow.Day}");
+
+                _logger.LogInformation($"Building http request with url {uri}");
+                var httpRequest = new HttpRequestMessage
+                {
+                    RequestUri = uri,
+                    Method = new HttpMethod("GET")
+                };
+
+                try
+                {
+                    _logger.LogInformation("Sending request to Doppler Currency Api.");
+                    var httpResponse = await _httpClient.SendAsync(httpRequest).ConfigureAwait(false);
+
+                    if (!httpResponse.IsSuccessStatusCode)
+                        continue;
+
+                    var json = await httpResponse.Content.ReadAsStringAsync();
+
+                    var result = JsonConvert.DeserializeObject<dynamic>(json);
+                    var date = result["entity"]["date"].ToString();
+                    var sale = result["entity"]["saleValue"].ToString();
+                    var buy = result["entity"]["buyValue"].ToString();
+                    var currencyName = result["entity"]["currencyName"].ToString();
```

Comment:
I would prefer to avoid convert to strings.

The result of the service is well known, why not simply deserialize to a predefined class?

Reply:
Yes, you are right, I fixed in this commit https://github.com/MakingSense/doppler-jobs/pull/1/commits/7fc34e978ac69c14d3933b5be6a59cddb75ed4d3